pkg/cloud/services/compute/instance.go

```diff
@@ -382,38 +382,64 @@ func (s *Service) getSecurityGroups(securityGroupParams []infrav1.SecurityGroupP
 
 func (s *Service) getServerNetworks(networkParams []infrav1.NetworkParam) ([]infrav1.Network, error) {
 	var nets []infrav1.Network
+
+	addSubnet := func(netID, subnetID string) {
+		nets = append(nets, infrav1.Network{
+			ID: netID,
+			Subnet: &infrav1.Subnet{
+				ID: subnetID,
+			},
+		})
+	}
+
+	addSubnets := func(networkParam infrav1.NetworkParam, netID string) error {
+		if len(networkParam.Subnets) == 0 && netID != "" {
+			addSubnet(netID, "")
+			return nil
+		}
+
+		for _, subnet := range networkParam.Subnets {
+			// Don't lookup subnet if it was specified explicitly by UUID
+			if subnet.UUID != "" {
+				addSubnet(netID, subnet.UUID)
```

Comment:
I think you should validate if `netID` is also not an empty string.

Reply:
Or don't we need the network ID here, as the subnet ID is sufficient?

Reply:
You are right. There are 2 ways we can go here:

* Fetch the subnet and therefore the network
* Return a validation error

I'm inclined to return the error as this is most efficiently fixed by the caller.

/hold

Reply:
Thinking out loud:

Q: How is this different to specifying a subnet filter without a network?
A: A UUID implies the caller already has the details of a specific resource. We are already trusting the caller if they specify a UUID (we use it without further validation): they will get a server create error instead if it was incorrect.

IOW if the caller specifies a network or subnet by UUID, we are explicitly delegating the validation of those UUIDs to the caller. If the caller specifies a filter we execute it for them, and let them know if something was wrong (i.e. it was empty).

Is that a good enough argument? If so, should we codify it in the docs?

Reply:
I think your assumption is correct. We should also document this preservation, I agree.
We don't have to make any code changes then. But it might be helpful to add a small comment. WDYT?

Reply:
Thanks. I will:

* update to return a validation here. If it's not sufficiently obvious from the validation error I'll also add a comment.
* add a unit test to cover this error.
* try to find some documentation to update.

On the last point, I want to make the following explicit in the docs:

* The controller will not validate an explicit network or subnet UUID.
* An invalid network or subnet UUID will result in a server create error.
* When specifying subnet UUID, network UUID must also be specified.